### DISCUSSION TITLE
TRANSFER USDT GASLESS FROM ETHEREUM WALLET AND GET THE GAS FEE SPONSORED BY BICONOMY

### DISCUSSION THREAD

**Discussion by subzz4u** [bcnmy/biconomy-client-sdk#547] [Q&A] [answered]
Hi Everyone, 

I have an use-case where i am trying to send USDT from an ethereum wallet and sponsor the transaction fee through biconomy.

Before that I have couple of questions

1. IS IT POSSIBLE TO TRANSFER USDT (ERC20) GAS LESS FROM AN ETHEREUM WALLET USING BICONOMY?

2. DOES BICONOMY SUPPORT GAS LESS TRANSACTIONS FOR ETHEREUM NETWOTK OR MAINNET CHAIN ? BECAUSE I TRIED BICONOMY HYPHEN , AND THERE GAS LESS IS DISABLED FOR ETHEREUM

3. IF I AM TRYING TO USE BICONOMY FIRST TIME ON MY WALLET DO I HAVE TO DO THE APPOVE TRANSACTION FIRST ?

4. CAN BICONOMY SPONSOR GAS FEE FOR APPROVE() TRANSACTIONS

`import { ethers, Signer } from "ethers";
import { createSmartAccountClient, PaymasterMode } from "@biconomy/account";

// Your configuration with private key and Biconomy API key
const config = {
    privateKey: "XXXX",
    biconomyPaymasterApiKey: "ly3EYQJuN.7e0d40b4-1e1b-4c45-844c-633c0a3d6343",
    bundlerUrl: "https://bundler.biconomy.io/api/v2/1/nJPK7B3ru.dd7f7861-190d-41bd-af80-6877f74b8f44", // <-- Read about this at https://docs.biconomy.io/dashboard#bundler-url
};

async function sendToBiconomy() {

    const usdtContractAddress = "0xdac17f958d2ee523a2206206994597c13d831ec7";  // Mainnet USDT contract address
    const recipientAddress = "XX";  // Replace with the recipient's address
    const amount = ethers.utils.parseUnits("1900", 6);
    const usdtContractABI = [
        {
            "constant": true,
            "inputs": [],
            "name": "name",
            "outputs": [
                {
                    "name": "",
                    "type": "string"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "_upgradedAddress",
                    "type": "address"
                }
            ],
            "name": "deprecate",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "_spender",
                    "type": "address"
                },
                {
                    "name": "_value",
                    "type": "uint256"
                }
            ],
            "name": "approve",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [],
            "name": "deprecated",
            "outputs": [
                {
                    "name": "",
                    "type": "bool"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "_evilUser",
                    "type": "address"
                }
            ],
            "name": "addBlackList",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [],
            "name": "totalSupply",
            "outputs": [
                {
                    "name": "",
                    "type": "uint256"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "_from",
                    "type": "address"
                },
                {
                    "name": "_to",
                    "type": "address"
                },
                {
                    "name": "_value",
                    "type": "uint256"
                }
            ],
            "name": "transferFrom",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [],
            "name": "upgradedAddress",
            "outputs": [
                {
                    "name": "",
                    "type": "address"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [
                {
                    "name": "",
                    "type": "address"
                }
            ],
            "name": "balances",
            "outputs": [
                {
                    "name": "",
                    "type": "uint256"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [],
            "name": "decimals",
            "outputs": [
                {
                    "name": "",
                    "type": "uint256"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [],
            "name": "maximumFee",
            "outputs": [
                {
                    "name": "",
                    "type": "uint256"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [],
            "name": "_totalSupply",
            "outputs": [
                {
                    "name": "",
                    "type": "uint256"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [],
            "name": "unpause",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [
                {
                    "name": "_maker",
                    "type": "address"
                }
            ],
            "name": "getBlackListStatus",
            "outputs": [
                {
                    "name": "",
                    "type": "bool"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [
                {
                    "name": "",
                    "type": "address"
                },
                {
                    "name": "",
                    "type": "address"
                }
            ],
            "name": "allowed",
            "outputs": [
                {
                    "name": "",
                    "type": "uint256"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [],
            "name": "paused",
            "outputs": [
                {
                    "name": "",
                    "type": "bool"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [
                {
                    "name": "who",
                    "type": "address"
                }
            ],
            "name": "balanceOf",
            "outputs": [
                {
                    "name": "",
                    "type": "uint256"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [],
            "name": "pause",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [],
            "name": "getOwner",
            "outputs": [
                {
                    "name": "",
                    "type": "address"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [],
            "name": "owner",
            "outputs": [
                {
                    "name": "",
                    "type": "address"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [],
            "name": "symbol",
            "outputs": [
                {
                    "name": "",
                    "type": "string"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "_to",
                    "type": "address"
                },
                {
                    "name": "_value",
                    "type": "uint256"
                }
            ],
            "name": "transfer",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "newBasisPoints",
                    "type": "uint256"
                },
                {
                    "name": "newMaxFee",
                    "type": "uint256"
                }
            ],
            "name": "setParams",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "amount",
                    "type": "uint256"
                }
            ],
            "name": "issue",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "amount",
                    "type": "uint256"
                }
            ],
            "name": "redeem",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [
                {
                    "name": "_owner",
                    "type": "address"
                },
                {
                    "name": "_spender",
                    "type": "address"
                }
            ],
            "name": "allowance",
            "outputs": [
                {
                    "name": "remaining",
                    "type": "uint256"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [],
            "name": "basisPointsRate",
            "outputs": [
                {
                    "name": "",
                    "type": "uint256"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [
                {
                    "name": "",
                    "type": "address"
                }
            ],
            "name": "isBlackListed",
            "outputs": [
                {
                    "name": "",
                    "type": "bool"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "_clearedUser",
                    "type": "address"
                }
            ],
            "name": "removeBlackList",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [],
            "name": "MAX_UINT",
            "outputs": [
                {
                    "name": "",
                    "type": "uint256"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "newOwner",
                    "type": "address"
                }
            ],
            "name": "transferOwnership",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "_blackListedUser",
                    "type": "address"
                }
            ],
            "name": "destroyBlackFunds",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "inputs": [
                {
                    "name": "_initialSupply",
                    "type": "uint256"
                },
                {
                    "name": "_name",
                    "type": "string"
                },
                {
                    "name": "_symbol",
                    "type": "string"
                },
                {
                    "name": "_decimals",
                    "type": "uint256"
                }
            ],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "constructor"
        },
        {
            "anonymous": false,
            "inputs": [
                {
                    "indexed": false,
                    "name": "amount",
                    "type": "uint256"
                }
            ],
            "name": "Issue",
            "type": "event"
        },
        {
            "anonymous": false,
            "inputs": [
                {
                    "indexed": false,
                    "name": "amount",
                    "type": "uint256"
                }
            ],
            "name": "Redeem",
            "type": "event"
        },
        {
            "anonymous": false,
            "inputs": [
                {
                    "indexed": false,
                    "name": "newAddress",
                    "type": "address"
                }
            ],
            "name": "Deprecate",
            "type": "event"
        },
        {
            "anonymous": false,
            "inputs": [
                {
                    "indexed": false,
                    "name": "feeBasisPoints",
                    "type": "uint256"
                },
                {
                    "indexed": false,
                    "name": "maxFee",
                    "type": "uint256"
                }
            ],
            "name": "Params",
            "type": "event"
        },
        {
            "anonymous": false,
            "inputs": [
                {
                    "indexed": false,
                    "name": "_blackListedUser",
                    "type": "address"
                },
                {
                    "indexed": false,
                    "name": "_balance",
                    "type": "uint256"
                }
            ],
            "name": "DestroyedBlackFunds",
            "type": "event"
        },
        {
            "anonymous": false,
            "inputs": [
                {
                    "indexed": false,
                    "name": "_user",
                    "type": "address"
                }
            ],
            "name": "AddedBlackList",
            "type": "event"
        },
        {
            "anonymous": false,
            "inputs": [
                {
                    "indexed": false,
                    "name": "_user",
                    "type": "address"
                }
            ],
            "name": "RemovedBlackList",
            "type": "event"
        },
        {
            "anonymous": false,
            "inputs": [
                {
                    "indexed": true,
                    "name": "owner",
                    "type": "address"
                },
                {
                    "indexed": true,
                    "name": "spender",
                    "type": "address"
                },
                {
                    "indexed": false,
                    "name": "value",
                    "type": "uint256"
                }
            ],
            "name": "Approval",
            "type": "event"
        },
        {
            "anonymous": false,
            "inputs": [
                {
                    "indexed": true,
                    "name": "from",
                    "type": "address"
                },
                {
                    "indexed": true,
                    "name": "to",
                    "type": "address"
                },
                {
                    "indexed": false,
                    "name": "value",
                    "type": "uint256"
                }
            ],
            "name": "Transfer",
            "type": "event"
        },
        {
            "anonymous": false,
            "inputs": [],
            "name": "Pause",
            "type": "event"
        },
        {
            "anonymous": false,
            "inputs": [],
            "name": "Unpause",
            "type": "event"
        }
    ];

    // Generate EOA from private key using ethers.js
    const provider = new ethers.providers.JsonRpcProvider("https://mainnet.infura.io/v3/4d55d37187df474b9d70ee4e649d1XXX");
    const signer = new ethers.Wallet("XXX", provider);

    const usdtContract = new ethers.Contract(usdtContractAddress, usdtContractABI, signer);

    // Create Biconomy Smart Account instance
    const smartWallet = await createSmartAccountClient({
        signer,
        biconomyPaymasterApiKey: config.biconomyPaymasterApiKey,
        bundlerUrl: config.bundlerUrl,
    });

    const saAddress = await smartWallet.getAccountAddress();
    console.log("SA Address", saAddress);

    const transactionData  = await usdtContract.populateTransaction.transfer(recipientAddress, amount);
    console.log(transactionData);

    // const signedTx = await signer.signTransaction(transactionData);
    // console.log(signedTx)

    const tx = {
        to : recipientAddress,
        data: transactionData.data
    }

    // Send the transaction and get the transaction hash
    const userOpResponse = await smartWallet.sendTransaction(tx, {
        paymasterServiceData: { mode: PaymasterMode.SPONSORED },
    });
    const { transactionHash } = await userOpResponse.waitForTxHash();
    console.log("Transaction Hash", transactionHash);
    const userOpReceipt = await userOpResponse.wait();
    if (userOpReceipt.success == "true") {
        console.log("UserOp receipt", userOpReceipt);
        console.log("Transaction receipt", userOpReceipt.receipt);
    }
}

sendToBiconomy();`

This is my code can some one tell me if I am making any mistakes or anything issue with my `code```

Thanks in advance :-)

**Comment by himanshugarg06** [comment]
1. No, the SDK that you are referring to requires smart accounts, so gasLess directly wont be possible on the EOA ethereum wallet.
2. Yes, the new SDK is live on ethereum mainnet and testnet
3. All functionalities will be managed by smart accounts and your EOA wallet will only be used as a signer. Please read more about smart accounts here https://docs.biconomy.io/
4. yes gas Fees can be sponsored for approve transactions also.

**Comment by himanshugarg06** [comment]
For each signer a smart account 0xABC will be created and you will require to have the USDT in the 0xABC smart account.

**Comment by subzz4u** [comment]
Hi @himanshugarg06  thanks for the quick response.  
I am a bit confused and I apologise for that. 

So for my case, if I have a normal ethereum wallet I will not be able to do gas less transaction? 

My USDT tokens are in my normal ethereum wallet. 

Is it possible to Convert my normal ethereum wallet to smart account? 

Else what should be the process flow for my case? I have a normal ethereum wallet with private key, secret phrase. I have some amount of USDT in that ethereum wallet. I want to transfer those USDT to a secure address by paying gas fees through biconomy. Please guide 

Waiting for your help sir. Thank you :-)

**Comment by subzz4u** [comment]
One update when I executed above code I got the transaction hash back with a success response. 
Gas fee got deducted from gas tank of my paymaster. But USDT did not get transferred to the destination address

**Comment by subzz4u** [comment]
Hi @himanshugarg06  it is a bit of urgent please help me !!!!

**Comment by himanshugarg06** [comment]
"So for my case, if I have a normal ethereum wallet I will not be able to do gas less transaction?"
NO

you first need to create a smart account with your EOA wallet and then transfer USDT from EOA to smart account as you normally would.
Then from there onwards you can use gasless or gas sponsorship for your smart account transactions.

**Comment by subzz4u** [comment]
Will Mexa SDK help in my case? 
So my question is  if I use mexa sdk

1. Usdt Contract does not natively inherit Eip 2771, will mexa sdk able to help me in this case

2. If yes, do I need to call the approve transaction first? And  can I send approval transaction through mexa sdk and sponsor gas fee? And what will be the sponsor address?

**Reply by tomarsachin2271** [reply, marked as answer]
From what i understood, you want to do gasless transaction to transfer USDT from EOA account.
Unfortunately this is not possible using mexa sdk as you correctly mentioned that USDT does not natively inherit EIP2771. This is the limitation of gasless transactions via EOA accounts.

Only way to transfer UDST from an account is if the account is SmartAccount (Smart Contract Wallet) and also to move USDT from EOA to a SmartAccount, user themselves have to pay the gas fee. There's not easy way to do this.
But once user move their USDT from EOA to a SmartAccount, they'll be able to move that USDT to anywhere in a gasless way using paymasters.

**Reply by teepoint67** [reply]
USDT Convert from to Ethereum

**Comment by himanshugarg06** [comment]
cc @divyaN73
